### PR TITLE
Added event handler for cauldron water level modification to prevent untrusted players interacting with a plot

### PR DESCRIPTION
**Pull request by ThisIsIPat**
Changes in the water level of a cauldron are now cancelled unless a player modifies it and that player has permissions to interact with the plot

Valid permissions:
``plots.admin.interact.road``
``plots.admin.interact.unowned``
``plots.admin.interact.other``
``Player is trusted or owner of the plot``

**Fixes https://github.com/IntellectualSites/PlotSquared/issues/3034**

### CODE REVIEW

**Comment by ThisIsIPat** [comment]
Note: It still extinguishes the player. It merely doesn't change the water level.

Another note: The water level is now also not modified by rain or, since 1.17, dripstone. This could possibly be implemented via a flag. If requested, the PR will be modified so that rain and dripstone will still modify a cauldrons water level.

File: Bukkit/src/main/java/com/plotsquared/bukkit/listener/BlockEventListener.java [hunk 1]
```diff
@@ -479,6 +480,48 @@ public void onBlockSpread(BlockSpreadEvent event) {
...
+        Location location = BukkitUtil.adapt(event.getBlock().getLocation());
+        PlotArea area = location.getPlotArea();
+        if (area == null) {
+            return;
+        }
+        Plot plot = area.getPlot(location);
+        // TODO Add flag to allow / deny 1.17 dripstone-based machines
+        // TODO Add flag to allow / deny modification by rain
+        if (entity instanceof Player) {
+            Player player = (Player) entity;
```
Comment (NotMyFault) [inline Bukkit/src/main/java/com/plotsquared/bukkit/listener/BlockEventListener.java:495]:
```suggestion
        if (entity instanceof Player player) {
```

**Review by NotMyFault — CHANGES_REQUESTED** [review]

**Comment by NotMyFault** [comment]
> Note: It still extinguishes the player. It merely doesn't change the water level.

Afaik the enum CauldronLevelChangeEvent.ChangeReason takes an extinguish cause determining whether the entity should be extinguished or not.

**Comment by ThisIsIPat** [comment]
Should cauldron modification by rain be disabled by default?

**Comment by NotMyFault** [comment]
> Should cauldron modification by rain be disabled by default?

I'd enable it. We don't cancel too many natural events on default if not really necessary, see liquid flow flags e.g.

**Review by NotMyFault — APPROVED** [review]
Thanks for the pr (again)!

**Comment by ThisIsIPat** [comment]
@NotMyFault Modified event handling so the event represents being cancelled by PlotSquared properly (Instead of having the new cauldron water level be the old one).

Requesting re-review.

**Review by NotMyFault — APPROVED** [review]

**Review by SirYwell — APPROVED** [review]

**Pull request merged**